cmake/KokkosConfigCommon.cmake.in

```diff
@@ -203,6 +204,573 @@ int main()
+#
+function(kokkos_get_directory_targets)
+  cmake_parse_arguments(INP "" "SEARCH_DIR" "FOUND_TARGETS;EXCLUDE_DIRS" ${ARGN})
+  set(targets "")
+  set(subdirs "")
+
+  if(${INP_SEARCH_DIR} IN_LIST INP_EXCLUDE_DIRS)
+    message(
+      DEBUG
+      "Kokkos excludes dir ${INP_SEARCH_DIR} from setting build options as it was marked to be excluded by `kokkos_exclude_from_setting_build_properties`"
+    )
+  else()
+    get_property(targets DIRECTORY ${INP_SEARCH_DIR} PROPERTY BUILDSYSTEM_TARGETS)
+    get_property(subdirs DIRECTORY ${INP_SEARCH_DIR} PROPERTY SUBDIRECTORIES)
+  endif()
+
+  # recursively visit subdirs
+  if(subdirs)
+    foreach(subdir ${subdirs})
+      kokkos_get_directory_targets(FOUND_TARGETS subdir_targets SEARCH_DIR ${subdir} EXCLUDE_DIRS ${INP_EXCLUDE_DIRS})
+      list(APPEND targets ${subdir_targets})
+    endforeach()
+  endif()
+
+  # push results out of recursive function
+  set(${INP_FOUND_TARGETS} ${targets} PARENT_SCOPE)
+endfunction()
+
+# this function searches a given directory recursively finding all targets that link to DEPENDENT_LIBRARY
+#
+#       GLOBAL         --> input: exclude everything from Kokkos setting build properties
+#       DIRECTORY      --> input: list of directories that should be excluded
+#
+function(kokkos_exclude_from_setting_build_properties)
+  cmake_parse_arguments(INP "GLOBAL" "" "DIRECTORY" ${ARGN})
+
+  if(INP_GLOBAL)
+    kokkos_exclude_from_setting_build_properties(DIRECTORY ${CMAKE_SOURCE_DIR})
+  else()
+    set(Kokkos_EXCLUDE_BUILD_DIRS
+        "${INP_DIRECTORY};${Kokkos_EXCLUDE_DIRS}"
+        CACHE
+          STRING
+          "List of directories Kokkos that are excluded from Kokkos setting build properties on source and target properties via a deferred call"
+          FORCE
+    )
+  endif()
+endfunction()
+
+# this function sets the source and target properties for all targets that link to a given library
+#
+#       LIBRARY                 --> input: name of the library downstream targets need to link to for this function to set the properties.
+#       USE_COMPILER_LAUNCHER   --> input: Command for a compiler launcher to use
+#       USE_LINKER_LAUNCHER     --> input: Command for a linker launcher to use
+#
+function(kokkos_set_dependent_library_properties)
+  cmake_parse_arguments(INP "" "" "LIBRARY;USE_COMPILER_LAUNCHER;USE_LINKER_LAUNCHER" ${ARGN})
+
+  if(INP_UNPARSED_ARGUMENTS)
+    message(
+      FATAL_ERROR "'kokkos_set_dependent_library_properties' has unrecognized arguments: ${INP_UNPARSED_ARGUMENTS}"
+    )
+  endif()
+
+  if(NOT INP_LIBRARY)
+    message(FATAL_ERROR "'kokkos_set_dependent_library_properties' needs a LIBRARY argument")
+  endif()
+
+  kokkos_check_compile_dependency(
+    library_compile_dependency
+    library_transitive_compile_dependency
+    library_link_dependency
+    "Kokkos::kokkoscore;Kokkos::kokkoscontainers;Kokkos::kokkosalgorithms;Kokkos::kokkossimd;Kokkos::kokkos"
```

Comment:
Perhaps not hard-code these libs?

Reply:
For now these are the ones that need compiler and linker settings downstream. We could rework the logic to curate this list automatically but this can also be done later. Added a FIXME comment to track it